onadata/apps/api/tools.py

```diff
@@ -415,10 +417,58 @@ class TagForm(forms.Form):
 
         if tags:
             for tag in tags:
-                instance.instance.tags.add(tag)
+                instance.tags.add(tag)
             instance.save()
 
 
+def add_validation_status_to_instance(request, instance):
+    """
+    Saves instance validation status if it's valid (belong to XForm/Asset validation statuses)
+
+    :param request: REST framework's Request object
+    :param instance: Instance object
+    :return: Boolean
+    """
+    validation_status_uid = request.data.get("validation_status__uid")
+    success = False
+
+    # Payload must contain validation_status property.
+    if validation_status_uid:
+
+        validation_status = get_validation_status(
+            validation_status_uid,instance.asset, request.user.username)
+        if validation_status:
+            instance.validation_status = validation_status
+            try:
+                instance.save()
+                instance.parsed_instance.update_mongo()
+                success = True
+            except Exception as e:
+                logging.error("add_validation_status_to_instance: {}".format(str(e)))
```

Comment:
Some things could probably happen in `save()` and `update_mongo()` that deserve a 500 response status instead of a 400. I'll suggest a change in a minute.

One good thing is that KoBoCAT has [`ATOMIC_REQUESTS`](https://github.com/kobotoolbox/kobocat/blob/cc634200c0ff90a7842075076c00fc1fb6b7f79d/onadata/settings/kc_environ.py#L17-L18) set to `True`, so the Postgres should be rolled back `update_mongo()` raises an exception. (Or not, because we catch the exception? I need to check.)